src/MonitorProfiler/Utilities/TypeNameUtilities.cpp

```diff
@@ -0,0 +1,249 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "TypeNameUtilities.h"
+#include "corhlpr.h"
+
+TypeNameUtilities::TypeNameUtilities(ICorProfilerInfo12* profilerInfo) : _profilerInfo(profilerInfo)
+{
+}
+
+HRESULT TypeNameUtilities::CacheNames(FunctionID functionId, COR_PRF_FRAME_INFO frameInfo, NameCache& nameCache)
+{
+    std::shared_ptr<FunctionData> functionData;
+    if (!nameCache.GetFunctionData(functionId, functionData))
+    {
+        return GetFunctionInfo(nameCache, functionId, frameInfo);
+    }
+
+    return S_OK;
+}
+
+HRESULT TypeNameUtilities::GetFunctionInfo(NameCache& nameCache, FunctionID id, COR_PRF_FRAME_INFO frameInfo)
+{
+    if (id == 0)
+    {
+        return E_INVALIDARG;
+    }
+
+    ClassID classId = 0;
+    ModuleID moduleId = 0;
+    mdToken token = mdTokenNil;
+    ULONG32 typeArgsCount = 0;
+    ClassID typeArgs[32];
+    HRESULT hr;
+
+    IfFailRet(_profilerInfo->GetFunctionInfo2(id,
+        frameInfo,
+        &classId,
+        &moduleId,
+        &token,
+        sizeof(typeArgs) / sizeof(ClassID),
+        &typeArgsCount,
+        typeArgs));
+
+    ComPtr<IMetaDataImport> pIMDImport;
+    IfFailRet(_profilerInfo->GetModuleMetaData(moduleId,
+        ofRead,
+        IID_IMetaDataImport,
+        (IUnknown**)&pIMDImport));
+
+    //TODO Convert this to dynamically allocate the needed size.
+    WCHAR funcName[256];
+    mdTypeDef classToken = mdTypeDefNil;
+    IfFailRet(pIMDImport->GetMethodProps(token,
+        &classToken,
+        funcName,
+        256,
+        0,
+        0,
+        NULL,
+        NULL,
+        NULL,
+        NULL));
+
+    IfFailRet(GetModuleInfo(nameCache, moduleId));
+
+    nameCache.AddFunctionData(moduleId, id, tstring(funcName), classId, classToken, typeArgs, typeArgsCount);
+
+    // If the ClassID returned from GetFunctionInfo is 0, then the function is a shared generic function.
+    if (classId != 0)
+    {
+        IfFailRet(GetClassInfo(nameCache, classId));
+    }
+    else
+    {
+        IfFailRet(GetTypeDefName(nameCache, moduleId, classToken));
+    }
+
+    for (ULONG32 i = 0; i < typeArgsCount; i++)
+    {
+        if (typeArgs[i] != 0)
+        {
+            IfFailRet(GetClassInfo(nameCache, typeArgs[i]));
+        }
+    }
+
+    return S_OK;
+}
+
+HRESULT TypeNameUtilities::GetClassInfo(NameCache& nameCache, ClassID classId)
+{
+    std::shared_ptr<ClassData> classData;
+    if (nameCache.GetClassData(classId, classData))
+    {
+        return S_OK;
+    }
+
+    if (classId == 0)
+    {
+        return E_INVALIDARG;
+    }
+
+    ModuleID modId = 0;
+    mdTypeDef classToken = mdTokenNil;
+    ULONG32 nTypeArgs = 0;
```

Comment:
nit: Above, I believe we do the same type of thing, but we name it `typeArgsCount`. Assuming they're the same concept, let's make the names consistent for readability.